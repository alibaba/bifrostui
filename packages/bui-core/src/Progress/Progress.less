.bui-progress {
  --fill-color: var(--bui-color-primary);
  --trail-color: rgba(156, 156, 165, 0.2);
  --stroke-width: 8px;
  width: 100%;
  font-family: var(--bui-font-family);

  &-inner {
    position: relative;
    display: inline-block;
    width: 100%;
    overflow: hidden;
    vertical-align: middle;
    background-color: var(--trail-color);
    border-radius: 100px;
  }

  &-bg {
    position: relative;
    border-radius: 100px;
    background-color: var(--fill-color);
    height: var(--stroke-width);
  }
}
