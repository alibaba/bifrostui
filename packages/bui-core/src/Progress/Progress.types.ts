import { OverrideProps } from '@bifrostui/types';
import React from 'react';

export type ProgressStringGradients = { [percentage: string]: string };
type ProgressFromToGradients = { from: string; to: string };
export type ProgressGradient = { direction?: string } & (
  | ProgressStringGradients
  | ProgressFromToGradients
);

export type ProgressProps<
  D extends React.ElementType = 'div',
  P = {},
> = OverrideProps<
  {
    props: P & {
      /** 百分比 */
      percent?: number;
      /**
       * - 进度条的色彩，传入 object 时为渐变
       */
      strokeColor?: string | ProgressGradient;
      /** 未填充进度条的色彩 */
      trailColor?: string;
      /** 进度条线的宽度，单位 */
      strokeWidth?: number | string;
    };
    defaultComponent: D;
  },
  D
>;
