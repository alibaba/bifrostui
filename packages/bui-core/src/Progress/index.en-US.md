---
group: feedback
name: Progress bar
---

# Progress bar

Display the current progress of the operation.

## Code demonstration

### Basic usage

Set the operation progress through 'percentage'.

```tsx
import React from 'react';
import { Progress } from '@bifrostui/react';
export default () => {
  return <Progress percent={40} />;
};
```

### Specify line color

Set the line color through 'strokeColor', with the default value being '-- bui color primary'.

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" spacing="5px">
      <Progress percent={40} strokeColor="var(--bui-color-primary)" />
      <Progress percent={50} strokeColor="var(--bui-color-info)" />
      <Progress percent={30} strokeColor="var(--bui-color-success)" />
      <Progress percent={80} strokeColor="var(--bui-color-warning)" />
      <Progress percent={90} strokeColor="var(--bui-color-vip)" />
    </Stack>
  );
};
```

### Progress bar width

You can set the width of the progress bar through 'strokeWidth'.

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" alignItems="flex-start" spacing="10px">
      <Progress
        percent={60}
        strokeColor="var(--bui-color-warning)"
        strokeWidth={10}
      />
      <Progress
        percent={60}
        strokeColor="var(--bui-color-warning)"
        strokeWidth={15}
      />
    </Stack>
  );
};
```

### Custom progress bar gradient color

Encapsulation of linear gradient. Recommend only transmitting two colors.

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" alignItems="flex-start" spacing="10px">
      <Progress
        percent={60}
        strokeColor={{
          from: 'var(--bui-color-red)',
          to: 'var(--bui-color-red-light)',
        }}
      />
      <Progress
        percent={90}
        strokeColor={{
          '0%': 'var(--bui-color-orange)',
          '100%': 'var(--bui-color-orange-light)',
        }}
      />
    </Stack>
  );
};
```

### Customize the color of unfinished segments

Specify the color of unfinished segments through 'trailColor'.

```tsx
import React from 'react';
import { Progress } from '@bifrostui/react';
export default () => {
  return <Progress percent={60} trailColor="var(--bui-color-red-light)" />;
};
```

## API

| attribute   | explain                                                             | type                                                                                                             | Default value |
| ----------- | ------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- | ------------- |
| percent     | percentage                                                          | number                                                                                                           | -             |
| strokeColor | The color of the progress bar is gradient when passing in an object | string \|{ from: string; to: string; direction?: string } \|{ [percentage: string]: string; direction?: string } | -             |
| trailColor  | Color of unfilled progress bar                                      | string                                                                                                           | -             |
| strokeWidth | Width of progress bar                                               | number \|string                                                                                                  | -             |

## Style variables

| attribute       | explain                        | Default value            | global variable              |
| --------------- | ------------------------------ | ------------------------ | ---------------------------- |
| --fill-color    | Color of progress bar          | --bui-color-primary      | -                            |
| --trail-color   | Color of unfilled progress bar | rgba(156, 156, 165, 0.2) | --bui-progress-trail-color   |
| --stroke-width  | Fill in progress bar width     | 8px                      | --bui-progress-stroke-width  |
| --width         | Progress bar width             | 100%                     | --bui-progress-width         |
| --border-radius | Round corner of progress bar   | 100px                    | --bui-progress-border-radius |
