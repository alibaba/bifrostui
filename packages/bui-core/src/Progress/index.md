---
group: 反馈
name: Progress 进度条
---

# Progress 进度条

展示操作的当前进度。

## 代码演示

### 基础用法

通过`percent`设置操作进度。

```tsx
import React from 'react';
import { Progress } from '@bifrostui/react';
export default () => {
  return <Progress percent={40} />;
};
```

### 指定线条颜色

通过`strokeColor`设置线条颜色，其默认值为`--bui-color-primary`。

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" spacing="5px">
      <Progress percent={40} strokeColor="var(--bui-color-primary)" />
      <Progress percent={50} strokeColor="var(--bui-color-info)" />
      <Progress percent={30} strokeColor="var(--bui-color-success)" />
      <Progress percent={80} strokeColor="var(--bui-color-warning)" />
      <Progress percent={90} strokeColor="var(--bui-color-vip)" />
    </Stack>
  );
};
```

### 进度条宽度

可以通过`strokeWidth`设置进度条宽度。

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" alignItems="flex-start" spacing="10px">
      <Progress
        percent={60}
        strokeColor="var(--bui-color-warning)"
        strokeWidth={10}
      />
      <Progress
        percent={60}
        strokeColor="var(--bui-color-warning)"
        strokeWidth={15}
      />
    </Stack>
  );
};
```

### 自定义进度条渐变色

`linear-gradient`的封装。推荐只传两种颜色。

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" alignItems="flex-start" spacing="10px">
      <Progress
        percent={60}
        strokeColor={{
          from: 'var(--bui-color-red)',
          to: 'var(--bui-color-red-light)',
        }}
      />
      <Progress
        percent={90}
        strokeColor={{
          '0%': 'var(--bui-color-orange)',
          '100%': 'var(--bui-color-orange-light)',
        }}
      />
    </Stack>
  );
};
```

### 自定义未完成的分段的颜色

通过`trailColor`指定未完成的分段的颜色。

```tsx
import React from 'react';
import { Progress } from '@bifrostui/react';
export default () => {
  return <Progress percent={60} trailColor="var(--bui-color-red-light)" />;
};
```

## API

| 属性        | 说明                               | 类型                                                                                                               | 默认值 |
| ----------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------------------ | ------ |
| percent     | 百分比                             | number                                                                                                             | -      |
| strokeColor | 进度条的色彩，传入 object 时为渐变 | string \| { from: string; to: string; direction?: string } \| { [percentage: string]: string; direction?: string } | -      |
| trailColor  | 未填充进度条的色彩                 | string                                                                                                             | -      |
| strokeWidth | 进度条线的宽度                     | number \| string                                                                                                   | -      |

## 样式变量

| 属性           | 说明               | 默认值                   |
| -------------- | ------------------ | ------------------------ |
| --fill-color   | 进度条的色彩       | --bui-color-primary      |
| --trail-color  | 未填充进度条的色彩 | rgba(156, 156, 165, 0.2) |
| --stroke-width | 进度条宽度         | 8px                      |
