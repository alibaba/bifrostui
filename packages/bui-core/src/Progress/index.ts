export { default } from './Progress';
export { default as Progress } from './Progress';
export * from './Progress.types';
