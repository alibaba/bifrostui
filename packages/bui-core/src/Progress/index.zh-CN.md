---
group: 反馈
name: Progress 进度条
---

# Progress 进度条

展示操作的当前进度。

## 基础用法

通过`percent`设置操作进度。

```tsx
import React from 'react';
import { Progress } from '@bifrostui/react';
export default () => {
  return <Progress percent={40} />;
};
```

### 指定线条颜色

通过`strokeColor`设置线条颜色，其默认值为`--bui-color-primary`。

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" spacing="5px">
      <Progress percent={40} strokeColor="var(--bui-color-primary)" />
      <Progress percent={50} strokeColor="var(--bui-color-info)" />
      <Progress percent={30} strokeColor="var(--bui-color-success)" />
      <Progress percent={80} strokeColor="var(--bui-color-warning)" />
      <Progress percent={90} strokeColor="var(--bui-color-vip)" />
    </Stack>
  );
};
```

### 进度条宽度

可以通过`strokeWidth`设置进度条宽度。

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" alignItems="flex-start" spacing="10px">
      <Progress
        percent={60}
        strokeColor="var(--bui-color-warning)"
        strokeWidth={10}
      />
      <Progress
        percent={60}
        strokeColor="var(--bui-color-warning)"
        strokeWidth={15}
      />
    </Stack>
  );
};
```

### 自定义进度条渐变色

`linear-gradient`的封装。推荐只传两种颜色。

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" alignItems="flex-start" spacing="10px">
      <Progress
        percent={60}
        strokeColor={{
          from: 'var(--bui-color-primary)',
          to: 'var(--bui-color-primary-light)',
        }}
      />
      <Progress
        percent={90}
        strokeColor={{
          '0%': 'var(--bui-color-orange)',
          '100%': 'var(--bui-color-orange-light)',
        }}
      />
    </Stack>
  );
};
```

### 自定义未完成的分段的颜色

通过`trailColor`指定未完成的分段的颜色。

```tsx
import React from 'react';
import { Progress } from '@bifrostui/react';
export default () => {
  return <Progress percent={60} trailColor="var(--bui-color-primary-light)" />;
};
```

## 无障碍性 (Accessibility)

Progress 组件内置了完整的无障碍性支持，可以直接使用以下 ARIA 属性：

- `aria-label`: 为进度条提供可访问的标签
- `aria-describedby`: 关联描述进度条的元素
- `aria-hidden`: 控制进度条在无障碍树中的可见性
- `aria-details`: 提供进度条的详细信息
- `aria-busy`: 表示进度条是否处于忙碌状态
- `aria-readonly`: 表示进度条是否为只读
- `aria-required`: 表示进度条是否为必填
- `aria-valuenow`: 进度条的当前值（默认为percent值）
- `aria-valuemin`: 进度条的最小值（默认为0）
- `aria-valuemax`: 进度条的最大值（默认为100）
- `aria-valuetext`: 进度条的文本描述

使用示例：

```tsx
import React from 'react';
import { Progress, Stack } from '@bifrostui/react';
export default () => {
  return (
    <Stack direction="column" spacing="10px">
      <Progress percent={50} aria-label="文件下载进度" />
      <Progress
        percent={75}
        aria-label="数据加载进度"
        aria-valuetext="已完成75%"
      />
      <Progress percent={30} aria-describedby="progress-description" />
      <span id="progress-description" style={{ display: 'none' }}>
        这是进度条的详细描述信息
      </span>
    </Stack>
  );
};
```

## API

| 属性        | 说明                               | 类型                                                                                                               | 默认值 |
| ----------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------------------ | ------ |
| percent     | 百分比                             | number                                                                                                             | -      |
| strokeColor | 进度条的色彩，传入 object 时为渐变 | string \| { from: string; to: string; direction?: string } \| { [percentage: string]: string; direction?: string } | -      |
| trailColor  | 未填充进度条的色彩                 | string                                                                                                             | -      |
| strokeWidth | 进度条线的宽度                     | number \| string                                                                                                   | -      |

## 样式变量

| 全局变量                     | 说明           | 默认值                     |
| ---------------------------- | -------------- | -------------------------- |
| --bui-progress-fill-color    | 进度条填充颜色 | `var(--bui-color-primary)` |
| --bui-progress-trail-color   | 进度条背景颜色 | `rgba(156, 156, 165, 0.2)` |
| --bui-progress-stroke-width  | 进度条高度     | `8px`                      |
| --bui-progress-width         | 进度条宽度     | `100%`                     |
| --bui-progress-border-radius | 进度条圆角     | `100px`                    |
