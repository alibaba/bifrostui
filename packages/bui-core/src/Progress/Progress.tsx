import clsx from 'clsx';
import React from 'react';
import {
  ProgressGradient,
  ProgressProps,
  ProgressStringGradients,
} from './Progress.types';
import './Progress.less';

const prefixCls = 'bui-progress';

function validProgress(progress: number | undefined) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}
const sortGradient = (gradients: ProgressStringGradients) => {
  let tempArr = [];
  Object.keys(gradients).forEach((key) => {
    const formattedKey = parseFloat(key.replace(/%/g, ''));
    if (!Number.isNaN(formattedKey)) {
      tempArr.push({
        key: formattedKey,
        value: gradients[key],
      });
    }
  });
  tempArr = tempArr.sort((a, b) => a.key - b.key);
  return tempArr.map(({ key, value }) => `${value} ${key}%`).join(', ');
};
export const handleGradient = (strokeColor: ProgressGradient) => {
  const { from, to, direction = 'to right', ...rest } = strokeColor;
  if (Object.keys(rest).length !== 0) {
    const sortedGradients = sortGradient(rest as ProgressStringGradients);
    return {
      backgroundImage: `linear-gradient(${direction}, ${sortedGradients})`,
    };
  }
  return { backgroundImage: `linear-gradient(${direction}, ${from}, ${to})` };
};

const Progress = React.forwardRef<HTMLDivElement, ProgressProps>(
  (props, ref) => {
    const {
      className,
      percent,
      strokeWidth,
      strokeColor,
      trailColor,
      ...others
    } = props;

    const backgroundProps =
      strokeColor && typeof strokeColor !== 'string'
        ? handleGradient(strokeColor)
        : {
            background: strokeColor,
          };

    const percentStyle = {
      width: `${validProgress(percent)}%`,
      height: strokeWidth,
      ...backgroundProps,
    } as React.CSSProperties;

    const trailStyle = {
      background: trailColor || undefined,
    } as React.CSSProperties;

    return (
      <div className={clsx(prefixCls, className)} ref={ref} {...others}>
        <div className={`${prefixCls}-inner`} style={trailStyle}>
          <div className={`${prefixCls}-bg`} style={percentStyle} />
        </div>
      </div>
    );
  },
);

Progress.displayName = 'BuiProgress';

export default Progress;
