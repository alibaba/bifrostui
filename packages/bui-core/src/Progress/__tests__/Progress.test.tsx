import React from 'react';
import { isConformant, render, screen, userEvent } from 'testing';
import Progress from '../index';
import { handleGradient } from '../Progress';

describe('Progress', () => {
  const rootClass = 'bui-progress';

  isConformant({
    Component: Progress,
    displayName: 'BuiProgress',
    className: rootClass,
  });

  it('renders correctly', () => {
    render(
      <Progress percent={54} strokeColor="#582331" data-testid="progress" />,
    );
    expect(screen.getByTestId('progress')).toMatchSnapshot();
  });

  it('should render correctly with percent props', () => {
    const P1 = render(<Progress percent={55} />);
    const P2 = render(<Progress percent={101} />);
    expect(P1.container.querySelector('.bui-progress-bg')).toHaveStyle({
      width: '55%',
    });
    expect(P2.container.querySelector('.bui-progress-bg')).toHaveStyle({
      width: '100%',
    });
  });

  it('can customize the color of the progress bar', () => {
    const { container } = render(<Progress percent={55} strokeColor="red" />);
    expect(container.querySelector('.bui-progress-bg')).toHaveStyle({
      background: 'red',
    });
  });

  it('can customize the color of the progress bar as gradient', () => {
    expect(
      handleGradient({
        from: 'red',
        to: 'blue',
        direction: 'to right',
        '0%': 'green',
        '50%': 'yellow',
        '100%': 'purple',
      }).backgroundImage,
    ).toBe('linear-gradient(to right, green 0%, yellow 50%, purple 100%)');
    expect(
      handleGradient({ '0%': '#108ee9', '100%': '#87d068' }).backgroundImage,
    ).toBe('linear-gradient(to right, #108ee9 0%, #87d068 100%)');
    expect(
      handleGradient({ from: 'red', to: 'blue', direction: 'to right' })
        .backgroundImage,
    ).toBe('linear-gradient(to right, red, blue)');
  });

  it('can customize the width of the progress bar', () => {
    const { container } = render(<Progress percent={55} strokeWidth={15} />);
    expect(container.querySelector('.bui-progress-bg')).toHaveStyle({
      height: '15px',
    });
  });

  it('can trigger a function by being clicked', () => {
    const onClick = jest.fn();
    render(<Progress percent={55} data-testid="progress" onClick={onClick} />);
    userEvent.click(screen.getByTestId('progress'));
    expect(onClick).toHaveBeenCalled();
  });
});
