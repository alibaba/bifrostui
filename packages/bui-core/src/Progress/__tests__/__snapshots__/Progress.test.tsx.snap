// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress renders correctly 1`] = `
<div
  class="bui-progress"
  data-testid="progress"
>
  <div
    class="bui-progress-inner"
  >
    <div
      class="bui-progress-bg"
      style="width: 54%; background: rgb(88, 35, 49);"
    />
  </div>
</div>
`;
